### ISSUE TITLE
use_gpu

### ISSUE THREAD

**Issue by tischi** [mpicbg-csbd/stardist#75]
Hello,

In the training notebooks there is this line of code:

```
# Use OpenCL-based computations for data generator during training (requires 'gputools')
use_gpu = False and gputools_available()
```

We were wondering what it exactly means and what one is supposed to do there? 
Probably change `False` to `True` if appropriate?

**Comment by maweigert** [comment]
Hi @tischi,

Yes, that might be a bit confusing at first. 
As the data generator that feeds the network has to compute the star-convex distances for every label image, this can be often be the bottleneck of the whole training process (i.e. the pre-processing dominates the actual CNN step at every iteration). To address that, we have implemented a GPU accelerated version (via [gputools](https://github.com/maweigert/gputools)) that can be switched on via `use_gpu = True` (we didn't use it as default, as it needs OpenCL and `gputools` to be installed). 

**tldr:** If you have `gputools` installed, just switch it on (and leave some space for OpenCL on the device via `csbdeep.utils.tf.limit_gpu_memory` which is done automatically in our training notebooks)

Hope that helps !

M

**Comment by tischi** [comment]
> and leave some space for OpenCL on the device via csbdeep.utils.tf.limit_gpu_memory which is done automatically in our training notebooks

does that mean I have to do something or is it done automatically?

**Comment by maweigert** [comment]
If you plan to use the gpu during training patch generation, you need to 
1) Install [gputools](https://github.com/maweigert/gputools) 
2) Limit the gpu memory (at the beginning of the script, before the model is set up)
```python 
from csbdeep.utils.tf import limit_gpu_memory
limit_gpu_memory(0.8)  # or 0.9, depending on your (b|p)atch_size 
```
3) Set `use_gpu=True` when creating the stardist model `Config`

**Comment by constantinpape** [comment]
Hi @maweigert,

for some context: we are currently setting up a user-friendly workflow for users to train stardist and we stumbled over this line.
Thanks to your explanation I understand what the `use_gpu` variable does now, but I still find the line

```
# Use OpenCL-based computations for data generator during training (requires 'gputools')
use_gpu = False and gputools_available()
```
a bit confusing.

Specifically, why do you use `use_gpu = False and ...` which will always evaluate to `False` ? 
Wouldn't `use_gpu = gputools_available()` be sufficient?
I would assume that `gputools_available` returns `True` if the additional dependency is available and `False` otherwise.

One more point: the variable name `use_gpu` could maybe be renamed to `use_gpu_for_data_processing` or so; because now it implies that `use_gpu = False` means the gpu is not used at all (which I know is not the case, but doesn't become quite clear from the notebook).

Sorry for being rather nit-picky here, but if this confuses us it will likely be confusing for users as well.
In any case, we will probably adapt the notebook a bit for users and just make these changes there, but we were wondering why you were choosing to do it like this in the first case.

**Comment by maweigert** [comment]
> a bit confusing.

I agree, we should change it a some point (the naming you suggest indeed makes a bit more sense). 
  

> Specifically, why do you use `use_gpu = False and ...` which will always evaluate to `False` ?

This is to ensure that `gputools` is available, once the user decides to set `use_gpu`. Note, that this is a bit of a power-user setting (if you want to speed up training), since installing `gputools`/`pyopencl` can be a bit involved on some systems. This is why it is deactivated by default. 

Hope that clears it up a bit and thanks for the feedback!  

M

**Comment by constantinpape** [comment]
> This is to ensure that `gputools` is available, once the user decides to set `use_gpu`. Note, that this is a bit of a power-user setting (if you want to speed up training), since installing `gputools`/`pyopencl` can be a bit involved on some systems. This is why it is deactivated by default.

Ok, now I understand it. Thanks for the clarification.

**Comment by uschmidt83** [comment]
> 2. Limit the gpu memory (at the beginning of the script, before the model is set up)

Note that this automatically happens in our training notebooks (cell 10/11), hence no need to do it manually:

```python
if use_gpu:
    from csbdeep.utils.tf import limit_gpu_memory
    # adjust as necessary: limit GPU memory to be used by TensorFlow to leave some to OpenCL-based computations
    limit_gpu_memory(0.8)
```

**Comment by uschmidt83** [comment]
> I still find the line
> 
> ```
> # Use OpenCL-based computations for data generator during training (requires 'gputools')
> use_gpu = False and gputools_available()
> ```
> 
> a bit confusing.

> One more point: the variable name `use_gpu` could maybe be renamed to `use_gpu_for_data_processing` or so; 
> Sorry for being rather nit-picky here, but if this confuses us it will likely be confusing for users as well.

In retrospect, I agree that `use_opencl`or similar would've been a better name for this.

But to be honest, I find it confusing that you (as a power user) are confused by this.
Isn't a comment in the notebook and explanation in the docstring enough?

```python
from stardist.models import Config2D
>>> help(Config2D)
Help on class Config2D in module stardist.models.model2d:
...
 |  use_gpu : bool
 |      Indicate that the data generator should use OpenCL to do computations on the GPU.
```

**Comment by constantinpape** [comment]
> In retrospect, I agree that `use_opencl`or similar would've been a better name for this.
> 
> But to be honest, I find it confusing that you (as a power user) are confused by this.
> Isn't a comment in the notebook and explanation in the docstring enough?

Yeah, I think that's fine for now; but still, descriptive names help ;).

Anyway, my main source of confusion was not the variable name, but the `False and ...` construction.
I would have thought that one could just set `use_gpu = gputools_available()`.
With Martin's explanation I understand the reason for this now.
